src/main/java/co/cask/gcp/datastore/sink/RecordToEntityTransformer.java

```diff
@@ -268,77 +272,107 @@ private Key transformKeyLiteralToKey(Schema.Field field, KeyFactory keyFactory,
+        return StringValue.newBuilder(stringValue)
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case INT:
       case LONG:
-        Number longValue = record.get(fieldName);
-        outValue = LongValue.newBuilder(longValue.longValue())
+        Number longValue = getValue(record::get, fieldName, fieldType.toString(), Number.class);
+        return LongValue.newBuilder(longValue.longValue())
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case FLOAT:
       case DOUBLE:
-        Number doubleValue = record.get(fieldName);
-        outValue = DoubleValue.newBuilder(doubleValue.doubleValue())
+        Number doubleValue = getValue(record::get, fieldName, fieldType.toString(), Number.class);
+        return DoubleValue.newBuilder(doubleValue.doubleValue())
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case BOOLEAN:
-        outValue = BooleanValue.newBuilder(record.get(fieldName))
+        Boolean booleanValue = getValue(record::get, fieldName, fieldType.toString(), Boolean.class);
+        return BooleanValue.newBuilder(booleanValue)
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case BYTES:
-        byte[] byteArray = record.get(fieldName);
-        outValue = BlobValue.newBuilder(Blob.copyFrom(byteArray))
+        byte[] byteArray = getValue(record::get, fieldName, fieldType.toString(), byte[].class);
+        return BlobValue.newBuilder(Blob.copyFrom(byteArray))
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case RECORD:
-        StructuredRecord nestedRecord = record.get(fieldName);
+        StructuredRecord nestedRecord = getValue(record::get, fieldName, fieldType.toString(), StructuredRecord.class);
         FullEntity.Builder<IncompleteKey> nestedBuilder = FullEntity.newBuilder();
         Objects.requireNonNull(fieldSchema.getFields(), "Nested Schema fields cannot be empty").forEach(
           nestedField -> nestedBuilder.set(nestedField.getName(),
                                            convertToValue(nestedField.getName(), nestedField.getSchema(),
                                                           nestedRecord, excludeFromIndex)));
-        outValue = EntityValue.newBuilder(nestedBuilder.build())
+        return EntityValue.newBuilder(nestedBuilder.build())
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
+      case ARRAY:
+        Schema elementSchema = Schema.recordOf("arrayElementSchema",
+                                               Schema.Field.of(fieldName, fieldSchema.getComponentSchema()
+                                               ));
+        List<?> arrayValues = getValue(record::get, fieldName, fieldType.toString(), List.class);
```

Comment:
The underlying value can also be an array. See https://github.com/cdapio/hydrator-plugins/blob/develop/hydrator-common/src/main/java/co/cask/hydrator/common/RecordConverter.java#L63.

Reply:
According to Datastore rules array cannot contain another array (https://cloud.google.com/datastore/docs/concepts/entities#array). Added this validation checks for Source and Sink schemas + unit tests.

Reply:
I meant the object returned by StructuredRecord.get() can by an array, it is not necessarily a List.

Reply:
Thanks for the clarification, updated the code and added unit test.